iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java

```diff
@@ -243,117 +243,121 @@ private RaftClientReply writeLocallyWithRetry(RaftClientRequest request) throws
     }
 
     if (suggestedLeader != null) {
       TEndPoint leaderEndPoint = Utils.fromRaftPeerAddressToTEndPoint(suggestedLeader.getAddress());
       writeResult.setRedirectNode(new TEndPoint(leaderEndPoint.getIp(), leaderEndPoint.getPort()));
     }
-    return ConsensusWriteResponse.newBuilder().setStatus(writeResult).build();
+    return writeResult;
   }
 
-  /** Read directly from LOCAL COPY notice: May read stale data (not linearizable) */
+  /** Read directly from LOCAL COPY notice. */
   @Override
-  public ConsensusReadResponse read(
-      ConsensusGroupId consensusGroupId, IConsensusRequest IConsensusRequest) {
-    RaftGroupId groupId = Utils.fromConsensusGroupIdToRaftGroupId(consensusGroupId);
-    RaftGroup group = getGroupInfo(groupId);
+  public DataSet read(ConsensusGroupId groupId, IConsensusRequest request)
+      throws ConsensusException {
+    RaftGroupId raftGroupId = Utils.fromConsensusGroupIdToRaftGroupId(groupId);
+    RaftGroup group = getGroupInfo(raftGroupId);
     if (group == null || !group.getPeers().contains(myself)) {
-      return failedRead(new ConsensusGroupNotExistException(consensusGroupId));
+      throw new ConsensusGroupNotExistException(groupId);
     }
 
     final boolean isLinearizableRead =
-        !canServeStaleRead.computeIfAbsent(consensusGroupId, id -> new AtomicBoolean(false)).get();
+        !canServeStaleRead.computeIfAbsent(groupId, id -> new AtomicBoolean(false)).get();
 
     RaftClientReply reply;
     try {
-      reply = doRead(groupId, IConsensusRequest, isLinearizableRead);
+      reply = doRead(raftGroupId, request, isLinearizableRead);
       // allow stale read if current linearizable read returns successfully
       if (isLinearizableRead) {
-        canServeStaleRead.get(consensusGroupId).set(true);
+        canServeStaleRead.get(groupId).set(true);
       }
-    } catch (Exception e) {
+    } catch (ReadException | ReadIndexException e) {
       if (isLinearizableRead) {
         // linearizable read failed. the RaftServer is recovering from Raft Log and cannot serve
         // read requests.
-        return failedRead(new RatisUnderRecoveryException(e));
+        throw new RatisUnderRecoveryException(e);
       } else {
-        return failedRead(new RatisRequestFailedException(e));
+        throw new RatisRequestFailedException(e);
       }
+    } catch (Exception e) {
+      throw new RatisRequestFailedException(e);
     }
-
     Message ret = reply.getMessage();
-    ResponseMessage readResponseMessage = (ResponseMessage) ret;
-    DataSet dataSet = (DataSet) readResponseMessage.getContentHolder();
-    return ConsensusReadResponse.newBuilder().setDataSet(dataSet).build();
+    org.apache.iotdb.consensus.ratis.ResponseMessage readResponseMessage =
+        (org.apache.iotdb.consensus.ratis.ResponseMessage) ret;
+    return Optional.ofNullable(readResponseMessage.getContentHolder())
+        .map(DataSet.class::cast)
+        .orElse(null);
```

Comment:
If the returned read response is null, we should force callers to handle this situation.
Or we can simply assert it is non-null.

Reply:
A null value here indicates a bug in the code, and avoiding an NPE isn't going to make much sense at this point, so I'll reverse the change